### PR TITLE
Add database support so comments are stored indefinitely 

### LINKED ISSUE #29: Add comment data to database in doPost() in DataServlet
- [x] Create `entity` using `comment` data entered by user in `<from>` from `comments.html`
- [x] Add `entity` to `database`

@billie-cat @f-camp

### LINKED ISSUE #30: Query comments from database in doGet() in DataServlet.java
- [x] `Query` database by `ASCENDING timestamp` order such that most recent comments are first
- [x] Loop over queried comments and add each `comment` object to `List<Comment>` for data to be parsed to `JSON`

@billie-cat @f-camp

### PR DESCRIPTION
**Pull request by jaysaleh**
This pull request adjust comments section such that when a `<form>` is submitted, the `comment` data is stored in a database. This allows for `comment` data to persist outside the life of the server.

Screenshots: https://screenshot.googleplex.com/6rt27rTjhso.png

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2019 Google LLC
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: This should be 2020

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.auto.value.AutoValue;
+
+/** Class used to create object to store comment data */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: Class comments are similar to method comments where you want to start with a verb. Also, we know its a class from line 21, so that can be omitted.

Here is more information on how to write class comments.
https://developers.google.com/style/api-reference-comments#classes_interfaces_structs

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+// TODO: Add database id feild for each comment.
+@AutoValue public abstract class Comment {
+  public abstract Long id();
+  public abstract String name();
+  public abstract String commentText();
+  public abstract Long timeStamp();
+
+  /** 
+   * Creates comment object
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]:
```suggestion
   * Creates a Comment.
```

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+   * @param name Name of author
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java:30]: Seems there are some params which are not explained.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.auto.value.AutoValue;
+
+/** Class used to create object to store comment data */
+// TODO: Add database id feild for each comment.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: This TODO seems completed.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,19 +14,81 @@
...
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns "Hello Jay Saleh!". TODO: modify this file to handle comments data */
+/** Servlet that writes an list of messages as a response. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
+  /** Creates a "comment" entity and stores in database */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: I would re-write this as: Creates and stores a {@code Comment} in the database.

@code tag will make it obvious this is a object/class.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -14,19 +14,81 @@
...
+    commentEntity.setProperty("timeStamp", timeStamp);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    response.sendRedirect("/html/comments.html");
+  }
+
+  /**
+   * Gets value from form in comments section of portfolio and returns it. If form is empty,
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I would rephrase this to be a bit more specific.

Example:
Returns value with {@code name} from the {@code request} form. If the {@code name} cannot be found, then returns {@code defaultValue}.

By tagging with @code you do not need to mention defaultValue was passed in.
Also, nothing in this code is guaranteeing that the request is of the comments section, so that can be omitted.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -14,19 +14,81 @@
...
+   * function will return defaultValue passed in.
+   * @param request Form sent by client
+   * @param name textArea to retrive information from
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: What is textArea? Should it have @code?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -14,19 +14,81 @@
...
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  @Override
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:82]: nit: @Override should be under the JavaDoc.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -14,19 +14,81 @@
...
+  /** Queries database for commenting in most recent time stamp order and writes comments as JSON to client */
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment").addSort("timeStamp", SortDirection.ASCENDING);
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: "Comment", "timestamp" and other fields seem to be used in multiple places. I would create a static final String for these at the top of the class.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -14,19 +14,81 @@
...
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  @Override
+  /** Queries database for commenting in most recent time stamp order and writes comments as JSON to client */
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:83]: Seems a bit weird/error prone to not be using request.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -14,19 +14,81 @@
...
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
+      return defaultValue;
+    }
+    return value;
+  }
+
+  @Override
+  /** Queries database for commenting in most recent time stamp order and writes comments as JSON to client */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: I would rephrase this a bit.

Example: Queries database for comments and writes them to {@code response} as JSON. Comments are written in order of most recent time stamp.

File: portfolio/src/main/webapp/script/comment-script.js [hunk 13]
```diff
@@ -12,12 +12,30 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/** @const commentHyphen used to space out comment text from author name */
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script/comment-script.js]: I believe this comment can be made into a summary. So you can just write: Spaces out comment text from author name.

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 1 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Done

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 3 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Done

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 4 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java:30]: Added the rest of them.

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 5 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 9 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:82]: Done

File: portfolio/src/main/webapp/script/comment-script.js (hunk 13 above)
Comment (jaysaleh) [reply portfolio/src/main/webapp/script/comment-script.js]: Done

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 2 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 12 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Removed override comments

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 11 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:83]: Removed override comments

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.auto.value.AutoValue;
+
+/** Creates a Comment for each comment written in portfolio. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: Sorry to comment on this again, but this class does not create comments, that is what the constructor does. This class stores data related to comments written in the porfolio.

/** Stores data for comments written in portfolio. */

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 15]
```diff
@@ -14,19 +14,91 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns "Hello Jay Saleh!". TODO: modify this file to handle comments data */
+/** Servlet that writes an list of messages as a response. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
/** Servlet that writes messages as a response. */
```

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 14 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 15 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed

File: portfolio/src/main/webapp/script/comment-script.js [hunk 16]
```diff
@@ -12,12 +12,29 @@
...
+    commentsListElement.appendChild(createListElement(comment.commentText + commentHyphen + comment.name));
+  }
+}
+
+/** 
+ * Creates and returns list element (<li>) containing {@code text} from comment.
+ */
+function createListElement(text) {
+  const liElement = document.createElement('li');
+  liElement.innerText = text;
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script/comment-script.js]: I am confused, wasn't this code already merged?

Comment (jaysaleh) [reply portfolio/src/main/webapp/script/comment-script.js]: Fixed

**Review by billie-cat — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 17]
```diff
@@ -17,18 +17,21 @@
...
    * Creates a Comment.
+   * @param id Id of object stored in database
    * @param name Name of author
-   * @param commentText Text written
+   * @param commentText Text author wrote
+   * @param timeStamp Time when author submitted comment
    */
-  public static Comment create(String name, String commentText) {
-    return new AutoValue_Comment(name, commentText);
+  public static Comment create(Long id, String name, String commentText, Long timeStamp) {
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: To avoid auto boxing can you use primitive long here?
```suggestion
  public static Comment create(long id, String name, String commentText, long timeStamp) {
```

**Review by f-camp — COMMENTED** [review]
Comments added.

File: portfolio/src/main/java/com/google/sps/data/Comment.java (hunk 17 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/data/Comment.java]: done

File: portfolio/src/main/webapp/script/comment-script.js (hunk 16 above)
Comment (jaysaleh) [reply portfolio/src/main/webapp/script/comment-script.js]: Fixed

**Review by billie-cat — APPROVED** [review]

**Review by f-camp — APPROVED** [review]
